app/master/build_request_handler.py

```diff
@@ -101,6 +101,11 @@ def _prepare_build_async(self, build, project_lock):
                                    log_msg='Build preparation loop is handling request for build {build_id}.')
             try:
                 build.prepare(self._subjob_calculator)
+                # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                if len(build.all_subjobs()) == 0:
+                    self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                    build.mark_started()
+                    build.perform_async_postbuild_tasks()
```

Comment:
+1 to adding something like `mark_completed()` or `finish()`. Performing postbuild tasks feels like an internal decision that the build should be making.

Also, ideally I'd rather us add the PREPARED -> FINISHED transition. Ideally we'd just call something like `finish()` on the build

I'm actually also fine with what you have right now. I still need to finish up a bunch of the work-in-progress notes I left around the FSM, which involves cleaning up the public interface anyway.